### PR TITLE
Port KafkaChannel to new bindings API

### LINKED ISSUE #976: Port KafkaChannel to cloudevents/sdk-go bindings
**Problem**
Now the `KafkaChannel` dispatcher uses the `Client` APIs of cloudevents/sdk-go

**Exit Criteria**
`KafkaChannel` dispatcher uses only the [bindings](https://github.com/cloudevents/sdk-go/tree/master/pkg/binding) APIs of cloudevents/sdk-go

**Time Estimate (optional):** 
1 week

**Additional context (optional)**
The new bindings API are designed to improve cpu/memory efficiency for transporting events.
Depends on #2671

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Fixes #976

### CODE REVIEW

**Comment by matzew** [comment]
cool, I will run some tests here

**Comment by slinkydeveloper** [comment]
/hold

I wish to fix the intermediate buffering that happens inside message receiver, which can be avoided for the kafka case

**Comment by slinkydeveloper** [comment]
Depends on https://github.com/knative/eventing/pull/2773

File: kafka/channel/pkg/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -63,6 +61,64 @@ type KafkaDispatcher struct {
...
+		args.Logger,
+		eventingchannels.ResolveMessageChannelFromHostHeader(dispatcher.getChannelReferenceFromHost))
+	if err != nil {
+		return nil, err
+	}
+
+	dispatcher.receiver = receiverFunc
+	dispatcher.setConfig(&multichannelfanout.Config{})
+	dispatcher.setHostToChannelMap(map[string]eventingchannels.ChannelReference{})
+	dispatcher.topicFunc = args.TopicFunc
```
Comment (lberk) [inline kafka/channel/pkg/dispatcher/dispatcher.go]: Do we need to set this again? will it change from the initial assignment on https://github.com/knative/eventing-contrib/pull/1039/files#diff-729c8026514d13072e43e868cd84cae5R93 ?

File: kafka/channel/pkg/dispatcher/dispatcher.go [hunk 2]
```diff
@@ -76,12 +132,15 @@ type KafkaDispatcherArgs struct {
...
 type consumerMessageHandler struct {
 	sub        subscription
-	dispatcher *eventingchannels.EventDispatcher
+	dispatcher *eventingchannels.MessageDispatcherImpl
 }
 
 func (c consumerMessageHandler) Handle(ctx context.Context, message *sarama.ConsumerMessage) (bool, error) {
-	event := fromKafkaMessage(ctx, message)
-	err := c.dispatcher.DispatchEventWithDelivery(ctx, *event, c.sub.SubscriberURI, c.sub.ReplyURI, &c.sub.Delivery)
+	m := protocolkafka.NewMessageFromConsumerMessage(message)
```
Comment (lberk) [inline kafka/channel/pkg/dispatcher/dispatcher.go]: Just a clean code nit; could we avoid one letter variable names where possible? Maybe something like `ceMessage` here given that `message` is an overloaded term wrt the sarama consumerMessage?

**Comment by lberk** [comment]
Outside of those two nits this works for me
/lgtm
/hold

**Comment by lberk** [comment]
@slinkydeveloper did you test these changes?  With the last commit the dispatcher doesn't forward any messages anymore for me.

Seems like the correct approach (for now) is to keep the variables assigned below the `NewMessageReciever` and remove the `topicFunc` assignment within the struct declaration.  Please make sure to remove the duplicate assignement (as it wasn't removed in 9fc5704

**Comment by slinkydeveloper** [comment]
> @slinkydeveloper did you test these changes? With the last commit the dispatcher doesn't forward any messages anymore for me.

Funny how the test pass...

**Comment by slinkydeveloper** [comment]
@lberk I've tested locally and it works :)

**Pull request merged**